python/ray/import_thread.py

```diff
@@ -0,0 +1,209 @@
+                    num_imports = self.redis_client.llen("Exports")
+                    assert num_imports >= num_imported
+                    for i in range(num_imported, num_imports):
+                        num_imported += 1
+                        key = self.redis_client.lindex("Exports", i)
+
+                        # Handle the driver case first.
+                        if self.mode != WORKER_MODE:
+                            if key.startswith(b"FunctionsToRun"):
+                                with log_span(
+                                        "ray:import_function_to_run",
+                                        worker=self.worker):
+                                    self.fetch_and_execute_function_to_run(key)
+                            # Continue because FunctionsToRun are the only
+                            # things that the driver should import.
+                            continue
+
+                        if key.startswith(b"RemoteFunction"):
+                            with log_span(
+                                    "ray:import_remote_function",
+                                    worker=self.worker):
+                                self.fetch_and_register_remote_function(key)
+                        elif key.startswith(b"FunctionsToRun"):
+                            with log_span(
+                                    "ray:import_function_to_run",
+                                    worker=self.worker):
+                                self.fetch_and_execute_function_to_run(key)
+                        elif key.startswith(b"ActorClass"):
+                            # Keep track of the fact that this actor class has
+                            # been exported so that we know it is safe to turn
+                            # this worker into an actor of that class.
+                            self.worker.imported_actor_classes.add(key)
+
+                        # TODO(rkn): We may need to bring back the case of
+                        # fetching actor classes here.
+                        else:
+                            raise Exception("This code should be unreachable.")
+        except redis.ConnectionError:
+            # When Redis terminates the listen call will throw a
+            # ConnectionError, which we catch here.
+            pass
+
+    def fetch_and_register_remote_function(self, key):
+        """Import a remote function."""
+        from ray.worker import FunctionExecutionInfo
+        (driver_id, function_id_str, function_name, serialized_function,
+         num_return_vals, module, resources,
+         max_calls) = self.redis_client.hmget(key, [
+             "driver_id", "function_id", "name", "function", "num_return_vals",
+             "module", "resources", "max_calls"
+         ])
+        function_id = ray.ObjectID(function_id_str)
+        function_name = function_name.decode("ascii")
+        max_calls = int(max_calls)
+        module = module.decode("ascii")
```

Comment:
If you rebase on the current master, all of the `decode("ascii")` lines should be replaced by `ray.utils.decode`. E.g., see https://github.com/ray-project/ray/pull/2306/commits/eaa11109f8e17e479b9a89198b3a94169bad7304.

This has better behavior in Python 2.